File: 2-hard/robot-movements/main.cpp
#include <iostream>

int countPaths(bool grid[][4], int currentX, int currentY)
{
    if (currentX < 0 || currentX > 3 || currentY < 0 || currentY > 3) {
        return 0;
    }

    if (grid[currentX][currentY]) {
        return 0;
    }

    if (currentX == 3 && currentY == 3) {
        return 1;
    }

    grid[currentX][currentY] = true;

    int result = 0;

    result += countPaths(grid, currentX+1, currentY);
    result += countPaths(grid, currentX,   currentY+1);
    result += countPaths(grid, currentX-1, currentY);
    result += countPaths(grid, currentX,   currentY-1);

    grid[currentX][currentY] = false;

    return result;
}

int main()
{
    bool grid[][4] = {
        {false, false, false, false},
        {false, false, false, false},
        {false, false, false, false},
        {false, false, false, false}
    };

    std::cout << countPaths(grid, 0, 0) << std::endl;
    return 0;
}

File: 2-hard/robot-movements/main.py
grid = [[False, False, False, False],
        [False, False, False, False],
        [False, False, False, False],
        [False, False, False, False]]

def count_paths(grid, current_x, current_y, end_x, end_y):
    if current_x < 0 or current_x > 3 or current_y < 0 or current_y > 3:
        return 0

    if grid[current_x][current_y]:
        return 0

    if current_x == end_x and current_y == end_y:
        return 1

    grid[current_x][current_y] = True

    result = 0

    result += count_paths(grid, current_x+1, current_y,   end_x, end_y)
    result += count_paths(grid, current_x,   current_y+1, end_x, end_y)
    result += count_paths(grid, current_x-1, current_y,   end_x, end_y)
    result += count_paths(grid, current_x,   current_y-1, end_x, end_y)

    grid[current_x][current_y] = False

    return result

print(count_paths(grid, 0, 0, 3, 3))
